### PR TITLE
Feat/add reset default prop selected filters component

### PR DESCRIPTION
**Pull request by mohdashraf010897**
**Description**: 
Added `resetToDefault` prop in SelectedFilters component, When `resetToDefault` is set to true, then it would set the filter value to the `defaultValue`.
The default value for `resetToDefault` would be false.

Libraries Affected 👍 
- RS: WEB
- RS: VUE

[Feature Request PR](https://github.com/appbaseio/reactivesearch/issues/1722)

[Notio Card](https://www.notion.so/appbase/ReactiveSearch-SelectedFilters-reset-to-default-value-99d4a8a722f14947ab41c33a18ee5312)

`web`
[Demo with resetToDefault prop ](https://www.loom.com/share/350c2e89b94b4ead8c74945efbccdb78)
[Demo w/o resetToDefault prop](https://www.loom.com/share/dee1baf641064a54ac15f5adf5049395)
[`MultiDropdownRange `Demo with resetToDefault prop ](https://www.loom.com/share/c3b5f3dd966542638df41e4d5acfae28)

`vue`
[Demo with resetToDefault prop  ](https://www.loom.com/share/28aa308a6e1c42789453221edcc57011)
[Demo w/o resetToDefault prop](https://www.loom.com/share/803641e1af2749b09b255777bf696dfb)
[`Range Input `Demo with resetToDefault prop  ](https://www.loom.com/share/655ee81ad8cc4266854225ce7479675e)

[Loom Demo: `Latest` Implementation](https://www.loom.com/share/bc69be97442144dbb801251ce999aac3)

[Loom Demo](https://www.loom.com/share/8eae8873854d465190dadd169ba6712c)

[Docs PR](https://github.com/appbaseio/Docs/pull/206)

### CODE REVIEW

File: packages/web/src/components/basic/SelectedFilters.js [hunk 1]
```diff
@@ -34,16 +34,35 @@ class SelectedFilters extends Component {
...
 	clearValues = () => {
-		const { onClear } = this.props;
-		this.props.clearValues();
+		const {
+			onClear, resetToDefault, componentProps, selectedValues,
+		} = this.props;
+		if (resetToDefault) {
+			Object.keys(selectedValues).map((component) => {
+				if (
+					componentProps[component].defaultValue
```
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js]: @mohdashraf010897 Add a `null` check before reading `defaultValue`.

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 2]
```diff
@@ -67,7 +69,23 @@ const SelectedFilters = {
 			this.$emit('clear', component, value);
 		},
 		clearValues() {
-			this.clearValuesAction();
+			const { resetToDefault, componentProps, selectedValues } = this;
+			if (resetToDefault) {
+				Object.keys(selectedValues).map(component => {
+					if (
+						componentProps[component].defaultValue
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: @mohdashraf010897 Add a `null` check before reading value. You can fix it at all the instances. Component shouldn't break if something is `undefined`.

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 3]
```diff
@@ -67,7 +69,23 @@ const SelectedFilters = {
...
+						&& !areArraysEqual(
+							selectedValues[component].value,
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: Add a null check before reading value.

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 4]
```diff
@@ -67,7 +69,23 @@ const SelectedFilters = {
 			this.$emit('clear', component, value);
 		},
 		clearValues() {
-			this.clearValuesAction();
+			const { resetToDefault, componentProps, selectedValues } = this;
+			if (resetToDefault) {
+				Object.keys(selectedValues).map(component => {
+					if (
+						componentProps[component].defaultValue
+						&& !areArraysEqual(
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: @mohdashraf010897 I would like to understand this logic. A loom video would help here.

**Review by bietkul — CHANGES_REQUESTED** [review]
@mohdashraf010897 There are some formatting changes. You can sync your branch with `next` branch and then fix the changes.

File: packages/vue/src/components/basic/SelectedFilters.jsx (hunk 4 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/basic/SelectedFilters.jsx]: logic updated

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 5]
```diff
@@ -67,7 +71,43 @@ const SelectedFilters = {
 			this.$emit('clear', component, value);
 		},
 		clearValues() {
-			this.clearValuesAction();
+			const { resetToDefault, componentProps, selectedValues } = this;
+			if (resetToDefault) {
```
Comment (larsdouweschuitema) [inline packages/vue/src/components/basic/SelectedFilters.jsx:73]: For readability purposes you can deconstruct these conditions into separate methods with a single purpose.

File: packages/vue/src/utils/index.js [hunk 6]
```diff
@@ -132,3 +132,30 @@ export const debounce = (method, delay) => {
 		method();
 	}, delay);
 };
+
+// to check whether arrays are equal in values
+// irrespective of the order of entries
+export const areArraysEqual = (arr1, arr2) => {
```
Comment (larsdouweschuitema) [inline packages/vue/src/utils/index.js]: For these types of utilities you might want to consider using an existing library to avoid extra maintenance.
Of course I understand if you have made the technical decision to have these types of functionalities on the library itself giving more freedom towards customization. However this functionality looks like a common one. And possible others too.

File: packages/web/src/components/basic/SelectedFilters.js [hunk 7]
```diff
@@ -6,7 +6,8 @@ import { withTheme } from 'emotion-theming';
 import { setValue, clearValues } from '@appbaseio/reactivecore/lib/actions';
 import { componentTypes, CLEAR_ALL } from '@appbaseio/reactivecore/lib/utils/constants';
 import types from '@appbaseio/reactivecore/lib/utils/types';
-import { getClassName, handleA11yAction } from '@appbaseio/reactivecore/lib/utils/helper';
+import { getClassName, handleA11yAction, isEqual } from '@appbaseio/reactivecore/lib/utils/helper';
+import XDate from 'xdate';
```
Comment (larsdouweschuitema) [inline packages/web/src/components/basic/SelectedFilters.js]: Not sure if you have compared library. But this one is also a nice one:
https://date-fns.org/

File: packages/web/examples/MultiDropdownList/src/index.js [hunk 8]
```diff
@@ -28,7 +28,7 @@ class Main extends Component {
 					</div>
 
 					<div className="col">
-						<SelectedFilters />
+						<SelectedFilters resetToDefault />
```
Comment (bietkul) [inline packages/web/examples/MultiDropdownList/src/index.js]: Remove this change

File: packages/web/src/components/range/RangeInput.js [hunk 9]
```diff
@@ -33,15 +33,6 @@ class RangeInput extends Component {
 		};
 	}
 
-	componentDidMount() {
```
Comment (bietkul) [inline packages/web/src/components/range/RangeInput.js]: @mohdashraf010897 Why did you remove this?

File: packages/web/src/components/basic/SelectedFilters.js [hunk 10]
```diff
@@ -42,8 +43,78 @@ class SelectedFilters extends Component {
 	};
 
 	clearValues = () => {
-		const { onClear } = this.props;
-		this.props.clearValues();
+		const {
```
Comment (larsdouweschuitema) [inline packages/web/src/components/basic/SelectedFilters.js]: Similar to my other comment you may wish to deconstruct this huge method into different methods.
This will also make the writing of tests easier and testability/maintainability in general easier.

File: packages/web/src/components/range/RangeInput.js [hunk 11]
```diff
@@ -214,8 +205,6 @@ RangeInput.propTypes = {
 	themePreset: types.themePreset,
 	componentId: types.stringRequired,
 	includeNullValues: types.bool,
-	enableAppbase: types.bool,
```
Comment (bietkul) [inline packages/web/src/components/range/RangeInput.js]: @mohdashraf010897 We should merge changes carefully

File: packages/web/src/components/range/RangeInput.js [hunk 12]
```diff
@@ -232,7 +221,6 @@ const mapStateToProps = (state, props) => ({
 	selectedValue: state.selectedValues[props.componentId]
 		? state.selectedValues[props.componentId].value
 		: null,
-	enableAppbase: state.config.enableAppbase,
```
Comment (bietkul) [inline packages/web/src/components/range/RangeInput.js]: Put it back

File: packages/web/src/utils/index.js [hunk 13]
```diff
@@ -169,3 +169,30 @@ export const getPopularSuggestionsComponent = (data = {}, props = {}) => {
 	}
 	return null;
 };
+
+// to check whether arrays are equal in values
+// irrespective of the order of entries
+export const areArraysEqual = (arr1, arr2) => {
```
Comment (larsdouweschuitema) [inline packages/web/src/utils/index.js]: You may wish to create a single utilities file shared among libraries instead of copy-pasta and ending up with duplicates.

File: packages/vue/src/components/range/MultiRange.jsx [hunk 14]
```diff
@@ -76,7 +75,10 @@ const MultiRange = {
 			} else {
 				const currentItems = props.data.filter(value => item.indexOf(value.label) !== -1);
 				currentValue = [...currentValue, ...currentItems];
-				selectedValues = { ...selectedValues, [item]: true };
+				selectedValues = {
+					...selectedValues,
+					[typeof item === 'object' ? item.label : item]: true,
```
Comment (bietkul) [inline packages/vue/src/components/range/MultiRange.jsx:80]: @mohdashraf010897 What is this change about?

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 15]
```diff
@@ -142,6 +182,7 @@ const SelectedFilters = {
 const mapStateToProps = state => ({
 	components: state.components,
 	selectedValues: state.selectedValues,
+	componentProps: state.props,
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: @mohdashraf010897 We don't want to re-render the component when `componentProps` would change. You can access the Redux store by using `$$store` property. Call `getState` on each demand, so you would get the fresh state.
alt img

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 16]
```diff
@@ -28,6 +31,7 @@ const SelectedFilters = {
 			return this.$scopedSlots.default({
 				components: this.components,
 				selectedValues: this.selectedValues,
+				componentProps: this.componentProps,
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: @mohdashraf010897 There is no need to provide it in default slot.

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 17]
```diff
@@ -67,7 +71,43 @@ const SelectedFilters = {
 			this.$emit('clear', component, value);
 		},
 		clearValues() {
-			this.clearValuesAction();
+			const { resetToDefault, componentProps, selectedValues } = this;
+			if (resetToDefault) {
+				let valueToSet;
+				Object.keys(selectedValues).map(component => {
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: @mohdashraf010897 Where are you using `mapped` value? Use `forEach` instead

File: packages/web/src/utils/index.js (hunk 13 above)
Comment (bietkul) [inline packages/web/src/utils/index.js]: @mohdashraf010897 Remove this because you're not using it.

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 18]
```diff
@@ -67,7 +71,43 @@ const SelectedFilters = {
...
+							componentTypes.dataSearch,
+							componentTypes.rangeSlider,
+							componentTypes.rangeInput,
+						].includes(componentProps[component].componentType)
+					) {
+						valueToSet = componentProps[component].defaultValue;
+					}
+
+					if (!isEqual(selectedValues[component].value, valueToSet)) {
+						this.setValue(component, valueToSet);
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx]: @mohdashraf010897 Instead of calling `setValue` for each component you should create a single action to reset the values for all components. The current solution is not performance efficient and could lead to re-rendering issues.

File: packages/web/src/components/basic/SelectedFilters.js (hunk 10 above)
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js]: @mohdashraf010897 Please refer to the comments on `Vue`.

File: packages/web/src/components/basic/SelectedFilters.js [hunk 19]
```diff
@@ -174,6 +245,7 @@ SelectedFilters.propTypes = {
 	title: types.title,
 	onChange: types.func,
 	componentProps: types.props,
```
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js]: @mohdashraf010897 I can see that `componentProps` is already being used here. I would recommend using the Context to read the store.
alt img

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/utils/index.js (hunk 13 above)
Comment (bietkul) [reply packages/web/src/utils/index.js]: It would be removed because of no usage.

File: packages/web/src/components/basic/SelectedFilters.js (hunk 7 above)
Comment (bietkul) [reply packages/web/src/components/basic/SelectedFilters.js]: Yes, we compare it. The `xdate` library is lighter than `date-fns`.

File: packages/vue/src/utils/index.js (hunk 6 above)
Comment (bietkul) [reply packages/vue/src/utils/index.js]: @mohdashraf010897 Please remove it because we're not using it.

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/vue/src/components/basic/SelectedFilters.jsx (hunk 5 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/basic/SelectedFilters.jsx:73]: Agreed! 👍  
A method named `clearValuesToDefault()` would be defined and used to clear values to default conditionally.

File: packages/web/examples/MultiDropdownList/src/index.js (hunk 8 above)
Comment (mohdashraf010897) [reply packages/web/examples/MultiDropdownList/src/index.js]: 👍

File: packages/web/src/components/range/RangeInput.js (hunk 9 above)
Comment (mohdashraf010897) [reply packages/web/src/components/range/RangeInput.js]: done

File: packages/web/src/components/range/RangeInput.js (hunk 11 above)
Comment (mohdashraf010897) [reply packages/web/src/components/range/RangeInput.js]: done 👍

File: packages/web/src/components/range/RangeInput.js (hunk 12 above)
Comment (mohdashraf010897) [reply packages/web/src/components/range/RangeInput.js]: 👍

File: packages/vue/src/components/range/MultiRange.jsx (hunk 14 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/range/MultiRange.jsx:80]: when values are selected from the range options they get passed in as objects and not simply 'strings', earlier implementation handles string values but. not objects and that's why I was not able to set values to default for this, here in the current updated implementation both data type objects and strings are accepted.
In short the edge case was unhandled.

File: packages/vue/src/components/basic/SelectedFilters.jsx (hunk 16 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/basic/SelectedFilters.jsx]: 👍

File: packages/vue/src/components/basic/SelectedFilters.jsx (hunk 17 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/basic/SelectedFilters.jsx]: 👍

File: packages/web/src/utils/index.js (hunk 13 above)
Comment (mohdashraf010897) [reply packages/web/src/utils/index.js]: 👍

File: packages/web/src/components/basic/SelectedFilters.js (hunk 10 above)
Comment (mohdashraf010897) [reply packages/web/src/components/basic/SelectedFilters.js]: Agreed! 👍
A method named clearValuesToDefault() would be defined and used to clear values to default conditionally.

Comment (mohdashraf010897) [reply packages/web/src/components/basic/SelectedFilters.js]: 👍

File: packages/web/src/components/basic/SelectedFilters.js (hunk 19 above)
Comment (mohdashraf010897) [reply packages/web/src/components/basic/SelectedFilters.js]: 👍  now using context

File: packages/vue/src/components/basic/SelectedFilters.jsx (hunk 15 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/basic/SelectedFilters.jsx]: 👍

**Comment by mohdashraf010897** [comment]
@bietkul [Demo: Updated Implementation](https://www.loom.com/share/bc69be97442144dbb801251ce999aac3)

File: packages/vue/src/components/basic/SelectedFilters.jsx [hunk 20]
```diff
@@ -30,6 +31,7 @@ const SelectedFilters = {
 				selectedValues: this.selectedValues,
 				clearValues: this.clearValues,
 				setValue: this.setValue,
+				resetValuesToDefault: this.resetValuesToDefault,
```
Comment (bietkul) [inline packages/vue/src/components/basic/SelectedFilters.jsx:34]: @mohdashraf010897 You would have to update the documentation for it,

File: packages/web/examples/CategorySearch/src/index.js [hunk 21]
```diff
@@ -27,11 +27,12 @@ const Main = () => (
 					enableRecentSearches
 					enablePopularSuggestions
 					size={5}
+					defaultValue={{ term: 'harry', category: '*' }}
 				/>
 			</div>
 
 			<div className="col">
-				<SelectedFilters />
+				<SelectedFilters resetToDefault />
```
Comment (bietkul) [inline packages/web/examples/CategorySearch/src/index.js]: @mohdashraf010897 Remove it

File: packages/web/src/components/basic/SelectedFilters.js [hunk 22]
```diff
@@ -3,16 +3,18 @@ import { jsx } from '@emotion/core';
 import React, { Component } from 'react';
 import { withTheme } from 'emotion-theming';
 
-import { setValue, clearValues } from '@appbaseio/reactivecore/lib/actions';
+import { setValue, clearValues, resetValuesToDefault } from '@appbaseio/reactivecore/lib/actions';
```
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js:5]: @mohdashraf010897 Why formatting issues?

File: packages/web/src/components/basic/SelectedFilters.js [hunk 23]
```diff
@@ -33,6 +35,8 @@ class SelectedFilters extends Component {
 		}
 	};
 
+	getStoreAccess = () => (this.context && this.context.storeState ? this.context.storeState : {});
```
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js]: @mohdashraf010897 It is wrong and not tested. Please fix it.

File: packages/web/src/components/basic/SelectedFilters.js [hunk 24]
```diff
@@ -103,19 +111,19 @@ class SelectedFilters extends Component {
 	};
 
 	// determines whether any filter has been applied regardless of `showFilter=false`
-	hasFilters = () =>
-		Object.keys(this.props.selectedValues)
+	hasFilters = () => {
+		const componentProps = this.getStoreAccess().props;
```
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js]: @mohdashraf010897 It seems wrong. You would get the redux store with `getState` method.

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/components/basic/SelectedFilters.js (hunk 24 above)
Comment (mohdashraf010897) [reply packages/web/src/components/basic/SelectedFilters.js]: `getStoreAccess = () => (this.context && this.context.storeState ? this.context.storeState : {});`

Have used context and defined a method to fetch it... Line 38

File: packages/vue/src/components/basic/SelectedFilters.jsx (hunk 20 above)
Comment (mohdashraf010897) [reply packages/vue/src/components/basic/SelectedFilters.jsx:34]: Screenshot 2021-06-28 at 11 12 29 PM

This is how docs have it until now, would need writeup and content to add to the doc...

File: packages/web/src/components/basic/SelectedFilters.js (hunk 23 above)
Comment (mohdashraf010897) [reply packages/web/src/components/basic/SelectedFilters.js]: 👍

File: packages/web/src/components/basic/SelectedFilters.js (hunk 24 above)
Comment (mohdashraf010897) [reply packages/web/src/components/basic/SelectedFilters.js]: ok

**Review by bietkul — APPROVED** [review]
@mohdashraf010897 Looks good now 👍

**Review by bietkul — APPROVED** [review]
